File: app/src/main/java/com/elementary/tasks/core/controller/LocationEvent.java
package com.elementary.tasks.core.controller;

import android.text.TextUtils;

import com.elementary.tasks.core.data.models.Reminder;
import com.elementary.tasks.core.services.EventJobService;
import com.elementary.tasks.core.services.GeolocationService;
import com.elementary.tasks.core.utils.Notifier;
import com.elementary.tasks.core.utils.RealmDb;
import com.elementary.tasks.core.utils.SuperUtil;
import com.elementary.tasks.core.utils.TimeCount;

import java.util.List;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

class LocationEvent extends EventManager {

    LocationEvent(Reminder reminder) {
        super(reminder);
    }

    @Override
    public boolean start() {
        getReminder().setActive(true);
        super.save();
        if (EventJobService.enablePositionDelay(getReminder().getUuId())) {
            return true;
        } else {
            SuperUtil.startGpsTracking(getContext());
            return true;
        }
    }

    @Override
    public boolean stop() {
        EventJobService.cancelReminder(getReminder().getUuId());
        getReminder().setActive(false);
        super.save();
        Notifier.hideNotification(getContext(), getReminder().getUniqueId());
        stopTracking(false);
        return true;
    }

    private void stopTracking(boolean isPaused) {
        List<Reminder> list = RealmDb.getInstance().getGpsReminders();
        if (list.size() == 0) {
            SuperUtil.stopService(getContext(), GeolocationService.class);
        }
        boolean hasActive = false;
        for (Reminder item : list) {
            if (isPaused) {
                if (item.getUniqueId() == getReminder().getUniqueId()) {
                    continue;
                }
                if (TextUtils.isEmpty(item.getEventTime()) || !TimeCount.isCurrent(item.getEventTime())) {
                    if (!item.isNotificationShown()) {
                        hasActive = true;
                        break;
                    }
                } else {
                    if (!item.isNotificationShown()) {
                        hasActive = true;
                        break;
                    }
                }
            } else {
                if (!item.isNotificationShown()) {
                    hasActive = true;
                    break;
                }
            }
        }
        if (!hasActive) {
            SuperUtil.stopService(getContext(), GeolocationService.class);
        }
    }

    @Override
    public boolean pause() {
        EventJobService.cancelReminder(getReminder().getUuId());
        stopTracking(true);
        return true;
    }

    @Override
    public boolean skip() {
        return false;
    }

    @Override
    public boolean resume() {
        if (getReminder().isActive()) {
            boolean b = EventJobService.enablePositionDelay(getReminder().getUuId());
            if (!b) SuperUtil.startGpsTracking(getContext());
        }
        return true;
    }

    @Override
    public boolean next() {
        return stop();
    }

    @Override
    public boolean onOff() {
        if (isActive()) {
            return stop();
        } else {
            getReminder().setLocked(false).setNotificationShown(false);
            super.save();
            return start();
        }
    }

    @Override
    public boolean isActive() {
        return getReminder().isActive();
    }

    @Override
    public boolean canSkip() {
        return false;
    }

    @Override
    public boolean isRepeatable() {
        return false;
    }

    @Override
    public void setDelay(int delay) {

    }

    @Override
    public long calculateTime(boolean isNew) {
        return TimeCount.getInstance(getContext()).generateDateTime(getReminder().getEventTime(), getReminder().getRepeatInterval());
    }
}

File: app/src/main/java/com/elementary/tasks/core/controller/LocationEvent.kt
package com.elementary.tasks.core.controller

import android.content.Context
import android.text.TextUtils
import com.github.naz013.appwidgets.AppWidgetUpdater
import com.github.naz013.feature.common.coroutine.invokeSuspend
import com.elementary.tasks.core.services.GeolocationService
import com.elementary.tasks.core.services.JobScheduler
import com.github.naz013.common.Module
import com.elementary.tasks.core.utils.Notifier
import com.elementary.tasks.core.utils.SuperUtil
import com.github.naz013.common.datetime.DateTimeManager
import com.elementary.tasks.core.utils.params.Prefs
import com.github.naz013.domain.Reminder
import com.github.naz013.repository.ReminderRepository
import org.threeten.bp.LocalDateTime

class LocationEvent(
  reminder: Reminder,
  private val reminderRepository: ReminderRepository,
  prefs: Prefs,
  private val context: Context,
  notifier: Notifier,
  private val jobScheduler: JobScheduler,
  appWidgetUpdater: AppWidgetUpdater,
  private val dateTimeManager: DateTimeManager
) : EventManager(reminder, reminderRepository, prefs, notifier, appWidgetUpdater) {

  override val isActive: Boolean
    get() = reminder.isActive

  override fun justStart() {
    if (Module.hasLocation(context)) {
      reminder.isActive = true
      reminder.isRemoved = false
      super.save()
      if (!jobScheduler.scheduleGpsDelay(reminder)) {
        SuperUtil.startGpsTracking(context)
      }
    }
  }

  override fun enable(): Boolean {
    return if (Module.hasLocation(context)) {
      reminder.isActive = true
      reminder.isRemoved = false
      super.save()
      if (jobScheduler.scheduleGpsDelay(reminder)) {
        true
      } else {
        SuperUtil.startGpsTracking(context)
        true
      }
    } else {
      disable()
      remove()
      false
    }
  }

  override fun disable(): Boolean {
    jobScheduler.cancelReminder(reminder.uniqueId)
    reminder.isActive = false
    if (prefs.moveCompleted) {
      reminder.isRemoved = true
    }
    super.save()
    notifier.cancel(reminder.uniqueId)
    stopTracking(false)
    return true
  }

  private fun stopTracking(isPaused: Boolean) {
    val list = invokeSuspend {
      reminderRepository.getAllTypes(
        active = true,
        removed = false,
        types = Reminder.gpsTypes()
      )
    }
    if (list.isEmpty()) {
      SuperUtil.stopService(context, GeolocationService::class.java)
    }
    var hasActive = false
    for (item in list) {
      if (isPaused) {
        if (item.uniqueId == reminder.uniqueId) {
          continue
        }
        if (TextUtils.isEmpty(item.eventTime) || !dateTimeManager.isCurrent(item.eventTime)) {
          if (!item.isNotificationShown) {
            hasActive = true
            break
          }
        } else {
          if (!item.isNotificationShown) {
            hasActive = true
            break
          }
        }
      } else {
        if (!item.isNotificationShown) {
          hasActive = true
          break
        }
      }
    }
    if (!hasActive) {
      SuperUtil.stopService(context, GeolocationService::class.java)
    }
  }

  override fun pause(): Boolean {
    jobScheduler.cancelReminder(reminder.uniqueId)
    notifier.cancel(reminder.uniqueId)
    stopTracking(true)
    return true
  }

  override fun skip(): Boolean {
    return false
  }

  override fun resume(): Boolean {
    if (reminder.isActive) {
      val b = jobScheduler.scheduleGpsDelay(reminder)
      if (!b) SuperUtil.startGpsTracking(context)
    }
    return true
  }

  override fun next(): Boolean {
    return disable()
  }

  override fun onOff(): Boolean {
    return if (isActive) {
      disable()
    } else {
      reminder.isLocked = false
      reminder.isNotificationShown = false
      super.save()
      enable()
    }
  }

  override fun canSkip(): Boolean {
    return false
  }

  override fun setDelay(delay: Int) {
  }

  override fun calculateTime(isNew: Boolean): LocalDateTime {
    return dateTimeManager.generateDateTime(reminder.eventTime, reminder.repeatInterval)
  }
}
